File: 2013-04-05/python/exercise05.py
### stairs 

step2D = MKPOL([[[0,0],[0,1.1],[0.6,0.5],[0.6,1.1]],
[[1,2,3,4]],None])
step3D = MAP([S1,S3,S2])(PROD([step2D,Q(2)]))

stair = STRUCT(NN(9)([step3D,T([1,3])([0.6,0.5])]))

### stair1
stair1 = T([1,2])([5.9,10.7])(stair)
### stair2
stair2 = T([1,2,3])([8,10.3,4.5])(stair)
### stair3
stair3 = T([1,2,3])([9.7,10.5,9.6])(stair)


building = STRUCT([pillars0,pillars1,pillars2,pillars3,floor0,floor1,floor2,floor3,floor4,south,west,est,north,stair1,stair2,stair3])

File: 2013-04-05/javascript/exercise05.js
T = function (dims) {
  dims = dims.map(function (dim) {
    return dim - 1;
  });

  return function (values) {
    return function (object) {
     return object.clone().translate(dims, values);
    };
  };
};
  
R = function (dims) {
  dims = dims.map(function (dim) {
    return dim - 1;
  });
   
  return function (angle) {
    return function (object) {
      return object.clone().rotate(dims, angle);
    };
  };
};
  
S = function (dims) {
  dims = dims.map(function (dim) {
    return dim - 1;
  });

  return function (values) {
    return function (object) {
      return object.clone().scale(dims, values);
    };
  };
};

S3 = S2;
S2 = S1;
S1 = S0;

CYLINDER = CYL_SURFACE;

GRID = SIMPLEX_GRID;

NN = REPLICA;

VIEW = DRAW;

var depth = 0.6;
var raiser = 0.5;
var step2D = SIMPLICIAL_COMPLEX([[0,0],[0,1.1],[depth,raiser],[depth,1.1]])([[0,2,1],[1,2,3]]);
var step3D = MAP([S0,S2,S1])(EXTRUDE([2])(step2D));
var stair = STRUCT(REPLICA(9)([step3D,T([1,3])([depth,raiser])]));


//stair1 floor1
var stair1 = T([1,2])([5.9,10.7])(stair);

//stair2 floor2
var stair2 = T([1,2,3])([8,10.3,4.5])(stair);

//stair3 floor3
var stair3 = T([1,2,3])([9.7,10.5,9.6])(stair);

var building = STRUCT([pillars0,pillars1,pillars2,pillars3,floor0,floor1,floor2,floor3,floor4,south,west,est,north,stair1,stair2,stair3]);
